### DISCUSSION TITLE
How to avoid memory leak in Angular 12 version?

### DISCUSSION THREAD

**Discussion by RamuDhanush** [community/community#145539] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

General

### Body

Hi,

We are currently using Angular version 12 for a large-scale application. Recently, we encountered a memory leak issue that results in high CPU usage and approximately 80,000 DOM nodes being created when navigating quickly between tabs. Garbage collection (GC) does not seem to clear these nodes, which causes the application to slow down.

Has anyone faced a similar issue? If so, please share any suggestions or solutions to resolve this.

Thanks,
Ramu

**Comment by DVSR1411** [comment]
Yes It's a very tricky problem. You can see my solution fits:
1. Update angular version as most of the memory leak issues are fixed in newer versions
2. Make sure to delete components and DOM elements after use and make sure all subscriptions (if any) are unsubscribed
3. Use async pipe whenever possible
4. Services should be used properly since they are shared resources
5. Use ChangeDetectionStrategy.OnPush to reduce unnecessary checks and increase efficiency
6. Use Chrome DevTools and Angular DevTools to monitor your app
7. Don't load too many modules and load modules only if you need them and don't load too often
8. Check Shadow DOM if it's causing any memory leaks
